### ISSUE TITLE
perl -P broken

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2270]
Migrated from [rt.perl.org#3588](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3588) (status was 'resolved')

Searchable as RT3588$

**Comment by p5pRT** [comment]
### From dcd@tc.fluke.com

#### Created by dcd@tc.fluke.com



cd t
\./perl harness \.//comp/cpp\.t
\.//comp/cpp\.\.\.\.\.\.\.\.\.FAILED before any test output arrived
FAILED\-\-1 test script could be run\, alas\-\-no output ever seen


\# perl 5\.6 rsynced from Jun 23 print ok 3 as expected\,
\( echo '\#define OK "ok 3\\n"' ; echo 'print OK;' \) | perl \-P \-
ok 3

\# but todays \(Fri Jul 28 17&#8203;:17&#8203;:47 PDT 2000\) code doesn't print ok
\#  fetched with
\#  rsync \-auvz rsync&#8203;://ftp\.linux\.activestate\.com/perl\-current/ perl\-current
\( echo '\#define OK "ok 3\\n"' ; echo 'print OK;' \) | \./perl \-P \-

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=critical

Site configuration information for perl v5.6.0:

Configured by dcd at Fri Jul 28 17:01:41 PDT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.0-test5, archname=i686-linux
    uname='linux dd 2.4.0-test5 #1 fri jul 28 11:04:28 pdt 2000 i686 '
    config_args='-Doptimize=-O3 -g -de -Dcf_email=dcd@tc.fluke.com'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O3 -g', gccversion=2.7.2.3, gccosandvers=linux
    cppflags='-DDEBUGGING -I/usr/local/include'
    ccflags ='-DDEBUGGING -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -ldbm -ldb -ldl -lm -lc
    libc=/lib/libc.so.5.4.44, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/local/lib/perl5/5.6.0/i686-linux
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/dcd
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/dcd/bin:/sbin:/usr/local/bin:/bin:/usr/bin:/usr/X11/bin:/usr/games:/usr/local/samba:/home/hobbes/tools/scripts:/home/hobbes/tools/linux:/usr0/hobbes/tools/scripts:/usr0/dcd/bin:/apps/general/bin:/usr/public
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @jhi

On Fri\, Jul 28\, 2000 at 05&#8203;:19&#8203;:16PM \-0700\, David Dyck wrote&#8203;:
> 
> This is a bug report for perl from dcd@&#8203;tc\.fluke\.com\,
> generated with the help of perlbug 1\.31 running under perl v5\.6\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> cd t
> \./perl harness \.//comp/cpp\.t
> \.//comp/cpp\.\.\.\.\.\.\.\.\.FAILED before any test output arrived
> FAILED\-\-1 test script could be run\, alas\-\-no output ever seen
> 
> 
> \# perl 5\.6 rsynced from Jun 23 print ok 3 as expected\,
> \( echo '\#define OK "ok 3\\n"' ; echo 'print OK;' \) | perl \-P \-
> ok 3
> 
> \# but todays \(Fri Jul 28 17&#8203;:17&#8203;:47 PDT 2000\) code doesn't print ok
> \#  fetched with
> \#  rsync \-auvz rsync&#8203;://ftp\.linux\.activestate\.com/perl\-current/ perl\-current
> \( echo '\#define OK "ok 3\\n"' ; echo 'print OK;' \) | \./perl \-P \-

That's strange\.\.\. firstly\, I do not off\-hand remember any patch lately
touching the \-P option\, at least not directly\.  Secondly\, for me the
test works in Linux\, and the configuration differences are&#8203;:

&nbsp;        you        me

Linux        2\.4\.0\-test5    2\.2\.16
gcc        2\.7\.2\.3        2\.95\.2 20000313 \(Debian GNU/Linux\)
lseeksize \*\)    4        8
libc        libc\.so\.5\.4\.44    libc\-2\.1\.3\.so
libs        \-lgdbm \-ldbm \-ldb \-ldl \-lm \-lc
&nbsp;                \-lnsl \-lndbm \-lgdbm \-ldb \-ldl \-lm \-lc \-lposix \-lcrypt \-lutil

\*\) We both have \-i686 and \-D\_LARGEFILE\_SOURCE \(and \-D\_FILE\_OFFSET\_BITS=64\)
&nbsp;   but our lseeksizes differ\.  Go figure\.

In other words&#8203;: I do not know what causes the breakage for you\.
But I suspect our gcc difference the most\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]


I've taken another look at this\, see my notes at the bottom

On Mon\, 31 Jul 2000\, Jarkko Hietaniemi wrote&#8203;:

> On Fri\, Jul 28\, 2000 at 05&#8203;:19&#8203;:16PM \-0700\, David Dyck wrote&#8203;:
> > 
> > This is a bug report for perl from dcd@&#8203;tc\.fluke\.com\,
> > generated with the help of perlbug 1\.31 running under perl v5\.6\.0\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> > 
> > cd t
> > \./perl harness \.//comp/cpp\.t
> > \.//comp/cpp\.\.\.\.\.\.\.\.\.FAILED before any test output arrived
> > FAILED\-\-1 test script could be run\, alas\-\-no output ever seen
> > 
> > 
> > \# perl 5\.6 rsynced from Jun 23 print ok 3 as expected\,
> > \( echo '\#define OK "ok 3\\n"' ; echo 'print OK;' \) | perl \-P \-
> > ok 3
> > 
> > \# but todays \(Fri Jul 28 17&#8203;:17&#8203;:47 PDT 2000\) code doesn't print ok
> > \#  fetched with
> > \#  rsync \-auvz rsync&#8203;://ftp\.linux\.activestate\.com/perl\-current/ perl\-current
> > \( echo '\#define OK "ok 3\\n"' ; echo 'print OK;' \) | \./perl \-P \-
> 
> That's strange\.\.\. firstly\, I do not off\-hand remember any patch lately
> touching the \-P option\, at least not directly\.  Secondly\, for me the
> test works in Linux\, and the configuration differences are&#8203;:
> 
>         you        me
> 
> Linux        2\.4\.0\-test5    2\.2\.16
> gcc        2\.7\.2\.3        2\.95\.2 20000313 \(Debian GNU/Linux\)
> lseeksize \*\)    4        8
> libc        libc\.so\.5\.4\.44    libc\-2\.1\.3\.so
> libs        \-lgdbm \-ldbm \-ldb \-ldl \-lm \-lc
>                 \-lnsl \-lndbm \-lgdbm \-ldb \-ldl \-lm \-lc \-lposix \-lcrypt \-lutil
> 
> \*\) We both have \-i686 and \-D\_LARGEFILE\_SOURCE \(and \-D\_FILE\_OFFSET\_BITS=64\)
>    but our lseeksizes differ\.  Go figure\.
> 
> In other words&#8203;: I do not know what causes the breakage for you\.
> But I suspect our gcc difference the most\.

When I reported this bug\, all I said that changes between Jun 23
and Jul 28 cause perl \-P to break\.

I took the effort to apply all the diffs before Jun 23 to
a fresh copy of 5\.6\.0\, and then applied each diff patch
in turn until my test case failed\.  I discovered
that diffs up to and including
6363\.gz and 6366\.gz \(Jul 11 14&#8203;:23 and Jul 11 14&#8203;:24\)
 \(\( Change 6363 by gsar@&#8203;auger on 2000/07/11 18&#8203;:49&#8203;:43
&nbsp;    Change 6366 by gsar@&#8203;auger on 2000/07/11 19&#8203;:11&#8203;:18 \)\)
must be applied to reproduce the failure\, and 
patch \-R backs out the failure and my test case works
again\.  \(I was able to test with just miniperl\, so it
wasn't too painful\)\.

I hope this will provide more information\, otherwise I guess I could
break these diffs into individual patches and apply each one
until the test case fails\.

I couldn't find the individual diff files on the
activestate ftp server\, but they used to be in
&nbsp;  ftp&#8203;://ftp\.linux\.activestate\.com/pub/staff/gsar/APC/diffs
and I used to get them from
&nbsp;  rsync \-auvz rsync&#8203;://ftp\.linux\.activestate\.com/perl\-diffs/ diffs

David Dyck

**Comment by p5pRT** [comment]
### From @gsar

On Tue\, 05 Sep 2000 17&#8203;:58&#8203;:59 PDT\, David Dyck wrote&#8203;:
>I couldn't find the individual diff files on the
>activestate ftp server\, but they used to be in
>  ftp&#8203;://ftp\.linux\.activestate\.com/pub/staff/gsar/APC/diffs

FYI\, diffs that went into a particular release can always be found
under the name of that release\.  So you'll find them at&#8203;:

&nbsp;   ftp&#8203;://ftp\.linux\.activestate\.com/pub/staff/gsar/APC/5\.7\.0/diffs
&nbsp;   ftp&#8203;://ftp\.linux\.activestate\.com/pub/staff/gsar/APC/5\.7\.1/diffs

etc\.  The APC/diffs symlink and the rsync\-able perl\-diffs just
point to the latest 5\.7\.x/diffs\, which means they move whenever
there's a new release\.


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]


Summary&#8203;:

I think that perl \-P was broken when the patch to toke\.c on
Tue Jul 11 14&#8203;:23&#8203;:09 2000 added the call to
&nbsp;  PerlIO\_tell\(PL\_rsfp\)==0\);
when perl was reading the piped output from cpp\.

Please see the end for more details\.

On Tue\, 5 Sep 2000\, David Dyck wrote&#8203;:

> I've taken another look at this\, see my notes at the bottom
> 
> On Mon\, 31 Jul 2000\, Jarkko Hietaniemi wrote&#8203;:
> 
> > On Fri\, Jul 28\, 2000 at 05&#8203;:19&#8203;:16PM \-0700\, David Dyck wrote&#8203;:
> > > 
> > > This is a bug report for perl from dcd@&#8203;tc\.fluke\.com\,
> > > generated with the help of perlbug 1\.31 running under perl v5\.6\.0\.
> > > 
> > > 
> > > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > > \[Please enter your report here\]
> > > 
> > > cd t
> > > \./perl harness \.//comp/cpp\.t
> > > \.//comp/cpp\.\.\.\.\.\.\.\.\.FAILED before any test output arrived
> > > FAILED\-\-1 test script could be run\, alas\-\-no output ever seen
> > > 
> > > 
> > > \# perl 5\.6 rsynced from Jun 23 print ok 3 as expected\,
> > > \( echo '\#define OK "ok 3\\n"' ; echo 'print OK;' \) | perl \-P \-
> > > ok 3
> > > 
> > > \# but todays \(Fri Jul 28 17&#8203;:17&#8203;:47 PDT 2000\) code doesn't print ok
> > > \#  fetched with
> > > \#  rsync \-auvz rsync&#8203;://ftp\.linux\.activestate\.com/perl\-current/ perl\-current
> > > \( echo '\#define OK "ok 3\\n"' ; echo 'print OK;' \) | \./perl \-P \-
> > 
> > That's strange\.\.\. firstly\, I do not off\-hand remember any patch lately
> > touching the \-P option\, at least not directly\.  Secondly\, for me the
> > test works in Linux\, and the configuration differences are&#8203;:
> > 
> >         you        me
> > 
> > Linux        2\.4\.0\-test5    2\.2\.16
> > gcc        2\.7\.2\.3        2\.95\.2 20000313 \(Debian GNU/Linux\)
> > lseeksize \*\)    4        8
> > libc        libc\.so\.5\.4\.44    libc\-2\.1\.3\.so
> > libs        \-lgdbm \-ldbm \-ldb \-ldl \-lm \-lc
> >                 \-lnsl \-lndbm \-lgdbm \-ldb \-ldl \-lm \-lc \-lposix \-lcrypt \-lutil
> > 
> > \*\) We both have \-i686 and \-D\_LARGEFILE\_SOURCE \(and \-D\_FILE\_OFFSET\_BITS=64\)
> >    but our lseeksizes differ\.  Go figure\.
> > 
> > In other words&#8203;: I do not know what causes the breakage for you\.
> > But I suspect our gcc difference the most\.
> 
> When I reported this bug\, all I said that changes between Jun 23
> and Jul 28 cause perl \-P to break\.
> 
> I took the effort to apply all the diffs before Jun 23 to
> a fresh copy of 5\.6\.0\, and then applied each diff patch
> in turn until my test case failed\.  I discovered
> that diffs up to and including
> 6363\.gz and 6366\.gz \(Jul 11 14&#8203;:23 and Jul 11 14&#8203;:24\)
>  \(\( Change 6363 by gsar@&#8203;auger on 2000/07/11 18&#8203;:49&#8203;:43
>     Change 6366 by gsar@&#8203;auger on 2000/07/11 19&#8203;:11&#8203;:18 \)\)
> must be applied to reproduce the failure\, and 
> patch \-R backs out the failure and my test case works
> again\.  \(I was able to test with just miniperl\, so it
> wasn't too painful\)\.
> 
> I hope this will provide more information\, otherwise I guess I could
> break these diffs into individual patches and apply each one
> until the test case fails\.

I broke patches 6363\.gz and 6366\.gz into individual file patches
and the following changes are what broke perl \-P for me
 \(it looks like the broken code gets an eof\, so it doesn't
&nbsp;  see any input \(or perhaps it only get the first line

&nbsp;    \#\!/usr/local/bin/perl \-P
&nbsp;    \#define OK "Ok 3\\n"
&nbsp;    BEGIN \{ print "BEGIN\\n"; \}
&nbsp;    print OK;
&nbsp;    BEGIN \{ print "END\\n"; \}

If I comment out the "&& \(PerlIO\_tell\(PL\_rsfp\)==0\)" in the line
&nbsp;    bof = PL\_rsfp && \(PerlIO\_tell\(PL\_rsfp\)==0\); /\* \*Before\* read\! \*/
then my test case starts to work\.  PerlIO\_tell was
returning ESPIPE\.
Of course\, this isn't the real fix\, but I hope this helps\.
My guess is that for my stdio library one should avoid
calling PerlIO\_tell on the file PL\_rsfp if PL\_rsfp is
the file handle that is know to be a pipe from cpp in perl \-P\.

==== //depot/perl/toke\.c\#244 \(text\) ====
Index&#8203;: perl/toke\.c
<details><summary>Inline Patch</summary>

```diff
--- perl/toke.c.~1~	Tue Jul 11 14:23:09 2000
+++ perl/toke.c	Tue Jul 11 14:23:09 2000
@@ -326,7 +326,7 @@
 }
 #endif
 
-#if 0
+#ifdef PERL_UTF16_FILTER
 STATIC I32
 S_utf16_textfilter(pTHX_ int idx, SV *sv, int maxlen)
 {
@@ -2490,6 +2490,8 @@
 	    goto retry;
 	}
 	do {
+        bool bof;
+        bof = PL_rsfp && (PerlIO_tell(PL_rsfp)==0); /* *Before* read! */
 	    if ((s = filter_gets(PL_linestr, PL_rsfp, 0)) == Nullch) {
 	      fake_eof:
 		if (PL_rsfp) {
@@ -2525,7 +2527,9 @@
 		    PL_bufend = SvPVX(PL_linestr) + SvCUR(PL_linestr);
 		    PL_doextract = FALSE;
 		}
-	    }
+	    } 
+        if (bof)
+            s = swallow_bom(s);
 	    incline(s);
 	} while (PL_doextract);
 	PL_oldoldbufptr = PL_oldbufptr = PL_bufptr = PL_linestart = s;
@@ -7407,3 +7411,55 @@
 	PerlIO_close(PL_rsfp);
     PL_rsfp = fp;
 }
+
+STATIC char*
+S_swallow_bom(pTHX_ char *s) {
+    STRLEN slen;
+    slen = SvCUR(PL_linestr);
+    switch (*s) {
+    case -1:       
+    if ((s[1] & 255) == 254) { 
+        /* UTF-16 little-endian */
+#ifdef PERL_UTF16_FILTER
+        U8 *news;
+#endif
+        s+=2;
+        if (*s == 0 && s[1] == 0)  /* UTF-32 little-endian */
+            Perl_croak(aTHX_ "Unsupported script encoding");
+#ifdef PERL_UTF16_FILTER
+        filter_add(S_utf16rev_textfilter, NULL);
+        New(898, news, (PL_bufend - s) * 3 / 2 + 1, U8);
+        PL_bufend = utf16_to_utf8((U16*)s, news, PL_bufend - s);
+        s = news;
+#else
+        Perl_croak(aTHX_ "Unsupported script encoding");
+#endif
+    }
+    break;
+
+    case -2:
+    if ((s[1] & 255) == 255) {   /* UTF-16 big-endian */
+#ifdef PERL_UTF16_FILTER
+        U8 *news;
+        filter_add(S_utf16_textfilter, NULL);
+        New(898, news, (PL_bufend - s) * 3 / 2 + 1, U8);
+        PL_bufend = utf16_to_utf8((U16*)s, news, PL_bufend - s);
+        s = news;
+#else
+        Perl_croak(aTHX_ "Unsupported script encoding");
+#endif
+   }
+   break;
+
+   case -17:
+   if ( slen>2 && (s[1] & 255) == 187 && (s[2] & 255) == 191) {
+        s+=3;                      /* UTF-8 */
+   }
+   break;
+   case 0:
+   if (slen > 3 && s[1] == 0 &&  /* UTF-32 big-endian */
+       s[2] & 255 == 254 && s[3] & 255 == 255)
+       Perl_croak(aTHX_ "Unsupported script encoding");
+} 
+return s;
+}
End of Patch.
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Tue\, 5 Sep 2000\, David Dyck wrote&#8203;:

> Summary&#8203;:
> 
> I think that perl \-P was broken when the patch to toke\.c on
> Tue Jul 11 14&#8203;:23&#8203;:09 2000 added the call to
>   PerlIO\_tell\(PL\_rsfp\)==0\);
> when perl was reading the piped output from cpp\.
> 
> Please see the end for more details\.
> 
> > On Mon\, 31 Jul 2000\, Jarkko Hietaniemi wrote&#8203;:
> > > That's strange\.\.\. firstly\, I do not off\-hand remember any patch lately
> > > touching the \-P option\, at least not directly\.  Secondly\, for me the
> > > test works in Linux\, and the configuration differences are&#8203;:
> > > 
> > >         you        me
> > > 
> > > Linux        2\.4\.0\-test5    2\.2\.16
> > > gcc        2\.7\.2\.3        2\.95\.2 20000313 \(Debian GNU/Linux\)
> > > lseeksize \*\)    4        8
> > > libc        libc\.so\.5\.4\.44    libc\-2\.1\.3\.so

I wrote a short c program that fails when linked with libc5
in a similar manner\, but works fine with libc6 \(glibc2\)

Am I the only linux perl 5\.7\.x user still using libc5?

David\.
